src/musicbrainz/web/coverartarchivelinkstask.cpp

```diff
@@ -0,0 +1,158 @@
+    networkAccessManager->setRedirectPolicy(QNetworkRequest::NoLessSafeRedirectPolicy);
+
+    const QNetworkRequest networkRequest = createNetworkRequest(m_albumReleaseId);
+
+    VERIFY_OR_DEBUG_ASSERT(url.isValid()) {
+        kLogger.warning() << "Invalid URL" << url;
+        return nullptr;
+    }
+    if (kLogger.traceEnabled()) {
+        kLogger.trace()
+                << "Get"
+                << url;
+    }
+    return networkAccessManager->get(networkRequest);
+}
+
+void CoverArtArchiveLinksTask::onFinished(
+        const network::JsonWebResponse& response) {
+    if (!response.isStatusCodeSuccess()) {
+        kLogger.warning()
+                << "Request failed with HTTP status code"
+                << response.statusCode();
+        emitFailed(response);
+        return;
+    }
+    VERIFY_OR_DEBUG_ASSERT(response.statusCode() == network::kHttpStatusCodeOk) {
+        kLogger.warning()
+                << "Unexpected HTTP status code"
+                << response.statusCode();
+        emitFailed(response);
+        return;
+    }
+    VERIFY_OR_DEBUG_ASSERT(response.content().isObject()) {
+        kLogger.warning()
+                << "Invalid JSON content"
+                << response.content();
+        emitFailed(response);
+        return;
+    }
+    const auto jsonObject = response.content().object();
+
+    if (jsonObject.isEmpty()) {
+        kLogger.warning()
+                << "Empty Json";
+        emitFailed(response);
+        return;
+    }
+
+    QList<QString> allUrls;
+    DEBUG_ASSERT(jsonObject.value(QLatin1String("images")).isArray());
+    const QJsonArray images = jsonObject.value(QLatin1String("images")).toArray();
+    for (const auto& image : images) {
+        DEBUG_ASSERT(image.isObject());
+        const auto imageObject = image.toObject();
+
+        const auto thumbnails = imageObject.value(QLatin1String("thumbnails")).toObject();
+        DEBUG_ASSERT(!thumbnails.isEmpty());
+
+        const auto smallThumbnailUrl = thumbnails.value(QLatin1String("small")).toString();
+        DEBUG_ASSERT(!smallThumbnailUrl.isNull());
+        allUrls.append(smallThumbnailUrl);
+
+        const auto largeThumbnailUrl = thumbnails.value(QLatin1String("large")).toString();
+        DEBUG_ASSERT(!largeThumbnailUrl.isNull());
+        allUrls.append(largeThumbnailUrl);
+
+        if (thumbnails.value(QLatin1String("1200")).toString() != nullptr) {
+            const auto largestThumbnailUrl = thumbnails.value(QLatin1String("1200")).toString();
+            allUrls.append(largestThumbnailUrl);
+        }
+
+        if (!(imageObject.value(QLatin1String("image")).isNull())) {
+            const auto highestResolutionImageUrl =
+                    imageObject.value(QLatin1String("image")).toString();
+            allUrls.append(highestResolutionImageUrl);
+        }
+
+        break;
+    }
+    emitSucceeded(allUrls);
+}
+
+void CoverArtArchiveLinksTask::emitSucceeded(
```

Comment:
Why is this a separate method?
IIUC it's not called from anywhere else so it can be part of `onFinished` above, no?

Reply:
Sorry I've missed that review.

So this is separate method because AFAIK it is on that method it checks if the signal is connected to Tagfetcher just in case of any missing connections. Correct me please If I am wrong. On other tasks such as [acoustidlookuptask.cpp](https://github.com/mixxxdj/mixxx/blob/8affbbb2c5e429b9aa8e6ea7410b65c8a50e46b1/src/musicbrainz/web/acoustidlookuptask.cpp#L213) and [musicbrainzrecordingtask.cpp](https://github.com/mixxxdj/mixxx/blob/8affbbb2c5e429b9aa8e6ea7410b65c8a50e46b1/src/musicbrainz/web/musicbrainzrecordingstask.cpp#L199). I just went with the flow. 

There were two missing methods in image task. First missing method was for succeeded. It was okay I guess. But the failed signal was missing (IDK If we needed it at all, because all the failings are handled in Links task, if there were no links image task was not starting to work, if there are links there was no failing in image task), now it is added with the method 👍

Reply:
I couldn't fully understand, sorry. 

Should we remove the methods? If yes, In other tasks some of the methods are also used once. So should we also remove them too?

Thanks for sharing the YAGNI principle by the way.

Reply:
I will keep it like this for the symmetry 👍 

Thank you.